src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
+
+A a2 = null;
+a2[y: x, x: x].X = ref i; // 2
+
+public class A
+{
+    public dynamic this[int x, int y] { get => throw null; }
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (4,36): error CS8373: The left-hand side of a ref assignment must be a ref variable.
+                // var a = new A() { [y: x, x: x] = { X = ref i } }; // 1
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "X").WithLocation(4, 36),
+                // (7,1): error CS8373: The left-hand side of a ref assignment must be a ref variable.
+                // a2[y: x, x: x].X = ref i; // 2
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "a2[y: x, x: x].X").WithLocation(7, 1)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_Escape()
+        {
+            var source = @"
+public class C
+{
+    public static R M1()
+    {
+        int x = 42;
+        var r = new R { field = ref x };
+        return r; // 1
+    }
+    public static R M2(ref int x)
+    {
+        var r = new R { field = ref x };
+        return r;
+    }
+    public static R M3()
+    {
+        R r = default;
+        {
+            int x = 42;
+            r = new R { field = ref x }; // 2
+        }
+        return r;
+    }
+    public static R M4()
+    {
+        R r = default;
+        int x = 42;
+        {
+            r = new R { field = ref x }; // 3
+        }
+        return r;
+    }
+    public static void M5(ref R r)
+    {
+        int x = 42;
+        r = new R { field = ref x }; // 4
+    }
+}
+
+public ref struct R
+{
+    public ref int field;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,16): error CS8352: Cannot use variable 'r' in this context because it may expose referenced variables outside of their declaration scope
+                //         return r; // 1
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "r").WithArguments("r").WithLocation(8, 16),
+                // (20,25): error CS8168: Cannot return local 'x' by reference because it is not a ref local
+                //             r = new R { field = ref x }; // 2
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "field = ref x").WithArguments("x").WithLocation(20, 25),
+                // (29,25): error CS8168: Cannot return local 'x' by reference because it is not a ref local
+                //             r = new R { field = ref x }; // 3
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "field = ref x").WithArguments("x").WithLocation(29, 25),
+                // (36,21): error CS8168: Cannot return local 'x' by reference because it is not a ref local
+                //         r = new R { field = ref x }; // 4
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "field = ref x").WithArguments("x").WithLocation(36, 21)
+                );
+
+            source = @"
```

Comment:
consider makign this a separate test and/or including a comment which indicates what the difference is between the two scenarios.